### PR TITLE
Resolve keyof and index operations instead of their targets.

### LINKED ISSUE #58752: [ts-5.5]Some types are excessively expanded in keyof and index access types
### 🔎 Search Terms

5.5rc regression type expanded declarations

### 🕗 Version & Regression Information

- This changed between versions 5.5 beta and 5.5 rc

### ⏯ Playground Link

[Workbench](https://www.typescriptlang.org/dev/bug-workbench/?target=2&ts=5.5.0-beta&ssl=6&ssc=50&pln=7&pc=50#code/PTAEAEBMFMGMBsCGAnRAXAlgewHYC5Q1kBXaAKBAjRQHNo0DoBnAJgAYBGAVjIrHABmGeNByIAttAIwEAOjRMysXEzShioALygARAA0dAbjJoAngAdooAIJbQAb1CICjgEagCOgG6J4pUFgCoK46ADSgOFhoAHJwzEwopgRmloHqoAC+GUoqaoh2js66Pn5WaYhhEVGxsPGJBBoZTkygyjiqvGTQAB7mWMhqbaoBHHYAFFgE1gDaAOSIswC6c65LAJRaAHwO2V29-YO5ASzjk6BjKdDla3MLixua2-bZPX0DrUdYAMynBIQWVyCiFu6y2O14rwOH3aaiwABZfqAANbQUxpMYzebrB5PXaUQTCUQSKSgcSIDA4eSKDDiN5qABUzVAkFAoAEyCw4lAs1kwBk8FmEP27yGaiC2nssilkAyQA)

### 💻 Code

```ts
// @declaration: true
// @target: es2015

// @filename: decl.ts
const u = "X";
type A = { a: { b : "value of b", notNecessary: typeof u }}
const a = { a: "value of a", notNecessary: u } as const


export const o1 = (o: A['a']['b']) => {}

export const o2 = (o: (typeof a)['a']) => {}
export const o3 = (o:  typeof a['a']) => {}

export const o4 = (o: keyof (A['a'])) => {}

// @filename: main.ts
import * as d  from './decl'

export const f = {...d}
```

### 🙁 Actual behavior

`A` and `(typeof a)` are expanded in declarations. 

### 🙂 Expected behavior

The parent type that uses `A` and `(typeof a)` should be expanded. Since we can't reuse the full node no sense in ballooning the type declaration.

### Additional information about the issue

Related to #58713

### PR DESCRIPTION
**Pull request by dragomirtitian**
Fixes #58752

### CODE REVIEW

File: src/compiler/utilities.ts [hunk 1]
```diff
@@ -11644,6 +11644,14 @@ export function unwrapParenthesizedExpression(o: Expression) {
     return o;
 }
 
+/** @internal */
+export function unwrapParenthesizedType(o: TypeNode) {
```
Comment (jakebailey) [inline src/compiler/utilities.ts:11648]: I'm surprised we don't already have this helper...

Comment (dragomirtitian) [reply src/compiler/utilities.ts:11648]: I was surprised too. But I could not find it if it exists. There is one that walks up.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -8605,7 +8606,54 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
...
+                        visitNodes(node.typeArguments, visitExistingNodeTreeSymbols, isTypeNode),
+                    );
+                }
+
+                const serializedName = serializeTypeName(context, node.exprName, /*isTypeOf*/ true);
+                if (serializedName) {
+                    return setTextRange(context, serializedName, node.exprName);
+                }
+            }
+            function tryVisitTypeReference(node: TypeReferenceNode): TypeNode | undefined {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:8656]:
```suggestion

            function tryVisitIndexAccess(node: IndexedAccessTypeNode): TypeNode | undefined {
                const resultObjectType = tryVisitSimpleTypeNode(node.objectType);
                if (resultObjectType === undefined) {
                    return undefined;
                }
                return factory.updateIndexedAccessTypeNode(node, resultObjectType, visitNode(node.indexType, visitExistingNodeTreeSymbols, isTypeNode)!);
            }

            function tryVisitKeyOf(node: TypeOperatorNode): TypeNode | undefined {
                Debug.assertEqual(node.operator, SyntaxKind.KeyOfKeyword);
                const type = tryVisitSimpleTypeNode(node.type);
                if (type === undefined) {
                    return undefined;
                }
                return factory.updateTypeOperatorNode(node, type);
            }

            function tryVisitTypeQuery(node: TypeQueryNode): TypeNode | undefined {
                const { introducesError, node: exprName } = trackExistingEntityName(node.exprName, context);
                if (!introducesError) {
                    return factory.updateTypeQueryNode(
                        node,
                        exprName,
                        visitNodes(node.typeArguments, visitExistingNodeTreeSymbols, isTypeNode),
                    );
                }

                const serializedName = serializeTypeName(context, node.exprName, /*isTypeOf*/ true);
                if (serializedName) {
                    return setTextRange(context, serializedName, node.exprName);
                }
            }

            function tryVisitTypeReference(node: TypeReferenceNode): TypeNode | undefined {

```

Comment (DanielRosenwasser) [reply src/compiler/checker.ts:8656]: It's just newlines between the declarations

**Comment by DanielRosenwasser** [comment]
Seems reasonable. Can you just address the style nits?

File: src/compiler/utilities.ts (hunk 1 above)
Comment (jakebailey) [reply src/compiler/utilities.ts:11648]: Ah, it's `skipTypeParentheses`.

Comment (dragomirtitian) [reply src/compiler/utilities.ts:11648]: Oh. Didn't think to look for that. I'll fix it tomorrow.

**Review by DanielRosenwasser — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
@typescript-bot cherry-pick to release-5.5

**Pull request merged**